File: skills/docforge/_shared/runtime/common/python/evidence_hash.py
"""Shared evidence hashing: raw git-blob hashes plus normalized and
line-range-scoped variants used to classify a cited source as fresh,
cosmetically drifted (whitespace/EOL-only, or the cited span is untouched),
or genuinely stale. Standard library only, no git dependency."""

from __future__ import annotations

import hashlib
import re
from pathlib import Path
from typing import Any

from runtime.common.python.provenance_frontmatter import BLOB

_LINE_SPLIT = re.compile(r"\r\n|\r|\n")
_RANGE_NUM = re.compile(r"^[1-9][0-9]*$")


def _decode_lines(content: bytes) -> list[str] | None:
    try:
        text = content.decode("utf-8")
    except UnicodeDecodeError:
        return None
    if text == "":
        return []
    lines = _LINE_SPLIT.split(text)
    if lines and lines[-1] == "" and text[-1:] in ("\n", "\r"):
        lines.pop()
    return lines


def raw_blob_hash(content: bytes) -> str:
    header = f"blob {len(content)}\0".encode("ascii")
    return hashlib.sha1(header + content).hexdigest()


def git_blob_for_path(path: Path) -> str | None:
    if not path.is_file():
        return None
    return raw_blob_hash(path.read_bytes())


def normalize_text_bytes(content: bytes) -> bytes | None:
    """CRLF/CR->LF, strip trailing whitespace per line, drop trailing blank
    lines at EOF. Deliberately does not touch comments, indentation, or
    internal blank-line runs -- conservative enough to stay safe across every
    language Docforge might cite."""
    lines = _decode_lines(content)
    if lines is None:
        return None
    trimmed = [line.rstrip(" \t") for line in lines]
    while trimmed and trimmed[-1] == "":
        trimmed.pop()
    if not trimmed:
        return b""
    return ("\n".join(trimmed) + "\n").encode("utf-8")


def normalized_blob_hash(content: bytes) -> str | None:
    normalized = normalize_text_bytes(content)
    return None if normalized is None else raw_blob_hash(normalized)


def line_count(content: bytes) -> int | None:
    lines = _decode_lines(content)
    return None if lines is None else len(lines)


def range_blob_hash(content: bytes, start: int, end: int) -> str | None:
    """Hash of the 1-indexed, inclusive line slice [start, end]."""
    lines = _decode_lines(content)
    if lines is None or start < 1 or end < start or end > len(lines):
        return None
    return raw_blob_hash("\n".join(lines[start - 1:end]).encode("utf-8"))


def _valid_range(evidence_range: Any, range_blob: Any) -> tuple[int, int] | None:
    if (
        isinstance(evidence_range, dict)
        and isinstance(evidence_range.get("start"), str) and _RANGE_NUM.match(evidence_range["start"])
        and isinstance(evidence_range.get("end"), str) and _RANGE_NUM.match(evidence_range["end"])
        and int(evidence_range["end"]) >= int(evidence_range["start"])
        and isinstance(range_blob, str) and BLOB.fullmatch(range_blob)
    ):
        return int(evidence_range["start"]), int(evidence_range["end"])
    return None


def classify_source(source: dict[str, Any], current_bytes: bytes | None) -> str:
    """Classify a provenance source given the CURRENT bytes of its cited path.

    Caller must validate the recorded `git_blob` is a well-formed 40-hex
    string before calling this -- a malformed/missing recorded blob is a
    NO_BLOB data-quality defect, never eligible for downgrade. Returns one of
    "missing", "fresh", "cosmetic", "stale", checked in that precedence:
    raw match (cheapest) -> range-scoped match (most specific) ->
    normalized match -> otherwise genuinely stale.
    """
    if current_bytes is None:
        return "missing"
    if raw_blob_hash(current_bytes) == source.get("git_blob"):
        return "fresh"
    span = _valid_range(source.get("evidence_range"), source.get("range_blob"))
    if span is not None:
        current_range = range_blob_hash(current_bytes, *span)
        if current_range is not None and current_range == source["range_blob"]:
            return "cosmetic"
    normalized = source.get("git_blob_normalized")
    if isinstance(normalized, str) and BLOB.fullmatch(normalized):
        current_normalized = normalized_blob_hash(current_bytes)
        if current_normalized is not None and current_normalized == normalized:
            return "cosmetic"
    return "stale"

File: skills/docforge/_shared/runtime/common/js/evidence_hash.js
"use strict";
/** Shared evidence hashing: raw git-blob hashes plus normalized and
 * line-range-scoped variants used to classify a cited source as fresh,
 * cosmetically drifted (whitespace/EOL-only, or the cited span is
 * untouched), or genuinely stale. Node built-ins only, no git dependency. */

const crypto = require("crypto");
const { BLOB } = require("./provenance_frontmatter.js");

const RANGE_NUM = /^[1-9][0-9]*$/;

function decodeLines(content) {
  let text;
  try {
    text = content.toString("utf8");
    if (Buffer.compare(Buffer.from(text, "utf8"), content) !== 0) return null;
  } catch {
    return null;
  }
  if (text === "") return [];
  const lines = text.split(/\r\n|\r|\n/);
  const last = text[text.length - 1];
  if (lines.length && lines[lines.length - 1] === "" && (last === "\n" || last === "\r")) {
    lines.pop();
  }
  return lines;
}

function rawBlobHash(content) {
  const header = Buffer.from(`blob ${content.length}\0`, "ascii");
  return crypto.createHash("sha1").update(Buffer.concat([header, content])).digest("hex");
}

function gitBlobForPath(fs, filePath) {
  if (!fs.existsSync(filePath) || !fs.statSync(filePath).isFile()) return null;
  return rawBlobHash(fs.readFileSync(filePath));
}

function normalizeTextBytes(content) {
  const lines = decodeLines(content);
  if (lines === null) return null;
  const trimmed = lines.map((line) => line.replace(/[ \t]+$/, ""));
  while (trimmed.length && trimmed[trimmed.length - 1] === "") trimmed.pop();
  if (!trimmed.length) return Buffer.alloc(0);
  return Buffer.from(`${trimmed.join("\n")}\n`, "utf8");
}

function normalizedBlobHash(content) {
  const normalized = normalizeTextBytes(content);
  return normalized === null ? null : rawBlobHash(normalized);
}

function lineCount(content) {
  const lines = decodeLines(content);
  return lines === null ? null : lines.length;
}

function rangeBlobHash(content, start, end) {
  const lines = decodeLines(content);
  if (lines === null || start < 1 || end < start || end > lines.length) return null;
  return rawBlobHash(Buffer.from(lines.slice(start - 1, end).join("\n"), "utf8"));
}

function validRange(evidenceRange, rangeBlob) {
  if (
    evidenceRange && typeof evidenceRange === "object" && !Array.isArray(evidenceRange)
    && typeof evidenceRange.start === "string" && RANGE_NUM.test(evidenceRange.start)
    && typeof evidenceRange.end === "string" && RANGE_NUM.test(evidenceRange.end)
    && Number(evidenceRange.end) >= Number(evidenceRange.start)
    && typeof rangeBlob === "string" && BLOB.test(rangeBlob)
  ) {
    return [Number(evidenceRange.start), Number(evidenceRange.end)];
  }
  return null;
}

function classifySource(source, currentBytes) {
  if (currentBytes === null || currentBytes === undefined) return "missing";
  if (rawBlobHash(currentBytes) === source.git_blob) return "fresh";
  const span = validRange(source.evidence_range, source.range_blob);
  if (span) {
    const currentRange = rangeBlobHash(currentBytes, span[0], span[1]);
    if (currentRange !== null && currentRange === source.range_blob) return "cosmetic";
  }
  const normalized = source.git_blob_normalized;
  if (typeof normalized === "string" && BLOB.test(normalized)) {
    const currentNormalized = normalizedBlobHash(currentBytes);
    if (currentNormalized !== null && currentNormalized === normalized) return "cosmetic";
  }
  return "stale";
}

module.exports = {
  rawBlobHash,
  gitBlobForPath,
  normalizeTextBytes,
  normalizedBlobHash,
  lineCount,
  rangeBlobHash,
  classifySource,
};
